Configurable test-on-borrow for pooled connections

lutovich opened the pull request:
Connections are pooled for both direct and routing driver. Each acquired session contains a socket connection which was acquired from the pool. It is possible for an idle pooled connection to become invalid while just resting in the pool. This could happen when connection is terminated after some timeout by a load balancer or some other network facility.

This PR introduces a configurable connection liveness check timeout. So freshly acquired connection, that has been idle in the pool for more than configurable timeout, will be validated by sending a RESET message. If it appears to be broken acquisition will retry until a valid connection is found.

lutovich committed:
Configurable test-on-borrow for pooled connections

Connections are pooled for both direct and routing driver. Each acquired
session contains a socket connection which was acquired from the pool.
It is possible for an idle pooled connection to become invalid while just
resting in the pool. This could happen when connection is terminated after
some timeout by a load balancer or some other network facility.

This commit introduces a configurable connection liveness check timeout.
So freshly acquired connection, that has been idle in the pool for more
than configurable timeout, will be validated by sending a RESET message.
If it appears to be broken acquisition will retry until a valid
connection is found.

zhenlineo commented on driver/src/main/java/org/neo4j/driver/v1/Config.java:
```diff
@@ -116,17 +123,14 @@ public int maxIdleConnectionPoolSize()
-     * method will always return <code>-1</code> and will be possibly removed in future.
      */
-    @Deprecated
     public long idleTimeBeforeConnectionTest()
```
There is no need to keep this old name. Pls pick a good name that better describes this config instead.

zhenlineo commented on driver/src/main/java/org/neo4j/driver/v1/Config.java:
```diff
@@ -256,30 +261,53 @@ public ConfigBuilder withMaxIdleSessions( int size )
          */
-        @Deprecated
-        public ConfigBuilder withSessionLivenessCheckTimeout( long timeout )
+        public ConfigBuilder withSessionLivenessCheckTimeout( long value, TimeUnit unit )
```
Same here, the name should be consistent with the field/method name in `Config`

lutovich committed:
Update license headers to 2017

zhenlineo commented on driver/src/main/java/org/neo4j/driver/internal/net/pooling/SocketConnectionPool.java:
```diff
@@ -106,34 +124,57 @@ private BlockingPooledConnectionQueue pool( BoltServerAddress address )
+
+        int acquisitionAttempt = 1;
+        PooledConnection connection = connectionQueue.acquire( supplier );
+        while ( !canBeAcquired( connection, acquisitionAttempt ) )
```
Suggesting to test queue.isEmpty() instead. And the while loop would be changed to something like
```
boolean done = false;
while(!done)
{
  if(connectionQueue.isEmpty())
  {
    done = true; // let `queue.acquire` for queue size + 1 times 
  }
  PooledConnection conn = connectionQueue.acquire(supplier);
  if(isHealthyConnection(conn))
  {
    return conn;
  }
}
// `isHealthyConnection` will not throw exception, so this is a final protection of NPE
throw new ClientException("Failed to establish connection with the server or some other better error message"); 


private boolean isHealthyConneciton(PooledConnection conn)
{
  if ( poolSettings.idleTimeBeforeConnectionTestConfigured() && hasBeenIdleForTooLong(conn) )
  {
    return connectionValidator.isConnected( connection );
  }
  return true;
}
```

lutovich replied:
Changed code to track when connection was created and then return it. So it'll retry until new connection is created and then return it to the user.

zhenlineo commented on driver/src/main/java/org/neo4j/driver/internal/net/pooling/SocketConnectionPool.java:
```diff
@@ -106,34 +124,57 @@ private BlockingPooledConnectionQueue pool( BoltServerAddress address )
 
-    @Override
-    public void close()
+    private boolean canBeAcquired( PooledConnection connection, int acquisitionAttempt )
```
This is not needed if we have `isHealthyConneciton` (see the previous comment)

zhenlineo commented on driver/src/main/java/org/neo4j/driver/internal/net/pooling/SocketConnectionPool.java:
```diff
@@ -68,26 +71,41 @@ public SocketConnectionPool( PoolSettings poolSettings, Connector connector, Clo
     {
         assertNotClosed();
+        BlockingPooledConnectionQueue connectionQueue = pool( address );
+        PooledConnection connection = acquireConnection( address, connectionQueue );
```
Not related to this PR but wonder if it would be better API if we let the pooled connection to hold a reference to connecitonValidator and let the pooledConnection to provide methods to test isConnected and isReusable.

lutovich replied:
It could also simplify things around these suppliers and consumers. Let's think about it.

zhenlineo commented on driver/src/main/java/org/neo4j/driver/v1/Config.java:
```diff
@@ -256,30 +261,53 @@ public ConfigBuilder withMaxIdleSessions( int size )
+            long idleTimeBeforeConnectionTestMillis = unit.toMillis( value );
+            if ( idleTimeBeforeConnectionTestMillis <= 0 )
+            {
+                throw new IllegalArgumentException( String.format(
```
Why not let the user to set to -1 (default value)?

lutovich committed:
Retry until connection is created

`SocketConnectionPool` can test connections during acquisition. This is done
when they have been idle in the pool for too long. This commit makes it retry
until new connection is created instead of `maxPoolSize`times.

lutovich committed:
Better naming of a config method

Renamed `withSessionLivenessCheckTimeout` to `withConnectionLivenessCheckTimeout`
because all other methods refer to connections instead of sessions and are
actually settings for the connection pool and not session pool.

lutovich committed:
Allow zero and negative values for `ConnectionLivenessCheckTimeout`

So feature can be easier turned on and off by users. Zero timeout means always
do test-on-borrow, negative timeout means never to test-on-borrow.

lutovich commented:
@zhenlineo comments addressed

zhenlineo commented on driver/src/main/java/org/neo4j/driver/v1/Config.java:
```diff
@@ -256,30 +261,47 @@ public ConfigBuilder withMaxIdleSessions( int size )
+        }
+
+        /**
+         * Pooled sessions that have been idle in the pool for longer than this timeout
```
Pls use `Pooled connections` instead as `pooled sessions` is a wrong terminology used before.

zhenlineo commented on driver/src/main/java/org/neo4j/driver/v1/Config.java:
```diff
@@ -116,17 +123,14 @@ public int maxIdleConnectionPoolSize()
     /**
-     * Pooled connections that have been unused for longer than this timeout will be tested before they are
-     * used again, to ensure they are still live.
+     * Pooled sessions that have been idle in the pool for longer than this timeout
```
pooled "connections"

zhenlineo reviewed (commented):
LGTM besides a minor doc change suggestion.

lutovich committed:
Fixed javadoc for connection liveness check timeout

Changed "session" to "connection" because config should only contain settings
for the connection pool. There is no such thing as session pool.
